### PR TITLE
[https://nvbugs/6293536][fix] Stage KV block offsets through a fresh host buffer

### PR DESCRIPTION
**Pull request by thorjohnsen**
## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed a data race condition in KV cache block offset handling during batch processing that could result in stale or incorrect data being used.

* **Tests**
  * Added regression test to prevent recurrence of the KV cache race condition.

## Description

Fixes a host-side data race (nvbug 6293536) that corrupts KV-cache block
indexing under the overlap scheduler, surfacing as CUDA memory corruption.

`KVCacheManager.copy_batch_block_offsets` shipped each iteration's block-offset
page tables to the device with an **asynchronous** H2D whose source was a
single, persistently-reused pinned host buffer (`host_kv_cache_block_offsets`)
overwritten **in place** every call. An async copy reads its source at
execution time, not enqueue time, so under the overlap scheduler — where the
CPU runs an iteration ahead — the next iteration's in-place overwrite could
clobber the source of the previous iteration's still-pending H2D. The attention
kernel then indexed another batch's physical blocks → corruption.

The race window is widened by v1 host KV offloading: `syncTransfers` stalls the
execution stream right in front of the offset H2D, so heavy offloading (driven
by concurrent prefill + primary-memory pressure) delays the copy long enough
for the host to overwrite the buffer first.

**Fix:** stage the offsets through a freshly-allocated pinned host buffer on
every call instead of reusing one in place. PyTorch's caching host allocator
holds a freed pinned buffer until the consuming async copy completes, so the
source can no longer be overwritten mid-flight — the same protection the
already-safe `kv_lens` / `block_ids_per_seq` staging paths rely on. The manager
attribute is reassigned to the current buffer so same-iteration CPU readers
(DSA sparse attention) and the speculative-decoding draft/target swap stay
coherent; `TrtllmAttentionMetadata.prepare` re-grabs the reference after the
copy.

This change is internal (no public API change).

## Test Coverage

`tests/unittest/_torch/executor/test_kv_block_offset_overlap_race.py` —
`test_copy_batch_block_offsets_survives_overlap_overwrite` drives the overlap
window deterministically with a `torch.cuda._sleep` stream stall (standing in
for the production `syncTransfers` offload stall), enqueues batch A's async H2D
behind the stall, then issues batch B before A's copy drains, and asserts batch
A's device offsets are not clobbered. The test fails on the pre-fix code and
passes with the fix. Existing `tests/unittest/_torch/executor/test_resource_manager.py`
continues to pass.

## PR Checklist

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

**Review by VALLIS-NERIA — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 1]
```diff
@@ -2070,6 +2070,27 @@ def pin_blocks(self, request_id: int):
...
+        # execution time, not enqueue time. With the overlap scheduler the CPU
+        # runs an iteration ahead, so reusing a single buffer let the next
+        # iteration's in-place overwrite clobber the source of this iteration's
+        # still-pending H2D -> the attention kernel indexed another batch's
+        # blocks (nvbug 6293536; the window is widened when host offloading
+        # stalls the execution stream in front of the H2D). A fresh allocation
+        # is held by PyTorch's caching host allocator until the consuming copy
+        # completes, matching the already-safe kv_lens / block_ids_per_seq
+        # staging.
+        host_block_offsets = torch.zeros(self.num_pools,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/resource_manager.py]: Should we Also fix this issue for KV cache manager v2?

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 2]
```diff
@@ -2070,6 +2070,27 @@ def pin_blocks(self, request_id: int):
...
+                                         num_seqs,
+                                         2,
+                                         self.max_blocks_per_seq,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/pyexecutor/resource_manager.py]: `max_blocks_per_seq` is a large number, so that size of host_block_offsets is large too. Instead of creating a new tensor every time, should we use double buffer instead? You can profile the buffer size to validate it.

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 3]
```diff
@@ -559,6 +559,11 @@ def prepare(self) -> None:
             self.kv_cache_manager.copy_batch_block_offsets(
                 self.kv_cache_block_offsets, self.request_ids, self.beam_width,
                 self.num_contexts, self.num_seqs)
+            # copy_batch_block_offsets stages through a fresh host buffer each
+            # iteration (nvbug 6293536 fix), so re-grab the current reference;
+            # the cached one from buffer allocation is now stale.
+            self.host_kv_cache_block_offsets = (
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: Should we also update `host_kv_cache_block_offsets` for draft KV cache manager in https://github.com/NVIDIA/TensorRT-LLM/blob/a862b7ca772090baf576764e379b8b4764d37b8c/tensorrt_llm/_torch/attention_backend/trtllm.py#L584

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py (hunk 1 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: We should fix this for v2 as well. Depending on what the fix is, we could include the v2 fix in this PR or fix it in a follow-up.

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py (hunk 2 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: Intuitively double buffering should be a little bit faster. I will look into it.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 3 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Probably should also be updated. I will look into it.

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py (hunk 2 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: Profiled this. Short version: the buffer can reach a few MB at long context, but it's **never larger than the persistent buffer this code already kept resident before the fix**, and the per-call allocation is effectively free in steady state — so I'd lean toward keeping fresh-per-call rather than adding a fixed double buffer.

**Size.** `host_block_offsets` is `[num_pools, num_seqs, 2, max_blocks_per_seq]` int32, where `num_seqs` is the *actual* batch (≤ `max_batch_size`). The pre-fix code already kept a persistent `[num_pools, max_batch_size, 2, max_blocks_per_seq]` buffer resident for the whole run, so the fresh buffer is equal-or-smaller — it doesn't raise the peak vs. the status quo. Measured (validated against real `KVCacheManager` builds; `max_blocks_per_seq = ceil(max_seq_len/tokens_per_block)`):

| config | b=32 | b=128 | b=256 |
|---|--:|--:|--:|
| 8K ctx | 0.06 MB | 0.25 MB | 0.50 MB |
| 128K ctx, block=64 | 0.50 MB | 2.0 MB | 4.0 MB |
| 128K ctx, block=32 | 1.0 MB | 4.0 MB | 8.0 MB |

So MB-scale only at 128K context + large batch; sub-MB otherwise.

**Cost.** In steady state the allocation is a PyTorch caching-host-allocator *cache hit*, not a `cudaHostAlloc`: ~25 µs warm and roughly size-independent (1 MB ≈ 16 MB). The first call per size bucket pays a cold `cudaHostAlloc` (ms-scale) but that's one-time/amortized. The full `copy_batch_block_offsets` call (alloc + C++ fill + async H2D enqueue) measured **median 11 µs** at 128K/batch64 — <0.1% of a (ms-scale) decode iteration.

**Why I'd skip the fixed double buffer.** The caching allocator already behaves as a self-retiring buffer pool — it recycles freed pinned blocks per size bucket, bounded to the overlap depth (~2). A fixed 2×full-capacity double buffer has a comparable footprint and the same ~11 µs, while re-introducing the wrap-onto-pending race this fix removes plus a buffer-count sizing question that's subtle under PP / spec decode. The keep-alive that protects the in-flight copy is free here because the consumer is a `cudaMemcpyAsync` out of pinned memory (the allocator records the event).

(For the v2 fix — #15607 — the same fresh-per-call approach works but needs an explicit CUDA-event guard, since the v2 consumer is a custom `cp.async` kernel the caching allocator doesn't auto-track.)

Happy to switch to a double buffer if you'd still prefer it, but based on the above fresh-per-call looks like the better tradeoff.

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py (hunk 1 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: Done — fixed in a follow-up PR: #15607.

The v2 manager has the same class of race but it needed a different fix, so it didn't fit cleanly into this PR. `KVCacheManagerV2.copy_batch_block_offsets` ships the offsets with `copy_batch_block_offsets_to_device`, a custom async `cp.async` gather kernel that reads its host inputs at execution time — and two of those inputs are reused in place across iterations: the persistent per-slot `host_kv_cache_block_offsets` (the C++ KV cache writes page indices straight into it, and a freed slot gets rebound to another request on reuse) and `copy_idx` (a slice of the IndexMapper's single persistent `copyIndex_` buffer).

The v1 trick doesn't transfer: the persistent buffer can't be reassigned (the C++ holds raw pointers into it via `set_base_page_index_buf`, and DSA reads it as the current page table), and a fresh per-call buffer alone isn't protected by the caching host allocator — that keep-alive is only recorded for `cudaMemcpyAsync` out of pinned memory, not for a custom kernel reading host memory directly. So #15607 snapshots the needed rows into a fresh pinned buffer + identity index and retires each buffer via an explicit CUDA event recorded after the kernel. Includes a v2 regression test (passes with the fix, fails without). Details in the PR description.

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 3 above)
Comment (thorjohnsen) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Looked into this — the draft path doesn't need the re-grab, and adding a symmetric assignment would actually create a dead attribute, so I addressed it with a clarifying comment (in #15607) instead.

The reason the target path re-grabs is that the metadata caches `self.host_kv_cache_block_offsets` at buffer-allocation time (`trtllm.py:324`), so it goes stale after `copy_batch_block_offsets` reassigns the manager's attribute. There's no analogous metadata-cached attribute for the draft side — only the device tensor `self.draft_kv_cache_block_offsets`, which is this metadata's own stable destination. The draft *host* offsets are read fresh through `draft_kv_cache_manager.host_kv_cache_block_offsets` by the draft/target swap in `speculative/interface.py`, which runs after `prepare()`, so consumers already observe the current buffer. Hence a comment rather than code.
